master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,609 @@
+    # Request.read(), there are many buffered readers in there and most likely multiple os.read()
+    # calls would occur and you'd be blocking anyway.
+    #
+    # So as an easy workaround, we set daemon=True and just exit the process if it's not done on
+    # time.
+    shipper = Shipper(master_url, token, cert_name, cert_file, logq, doneq, daemon=True)
+    shipper_timed_out = False
+
+    # Start the process or ship a special log message to the master why we couldn't.
+    try:
+        # Don't rely on Popen's standard line buffering; we want to do our own line buffering.
+        p = subprocess.Popen(
+            cmd, stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.PIPE, bufsize=0
+        )
+    except FileNotFoundError:
+        shipper.ship_special(f"FileNotFoundError executing {cmd}", metadata, emit_stdout_logs)
+        # 127 is the standard bash exit code for file-not-found.
+        return 127
+    except PermissionError:
+        # Unable to read or to execute the command.
+        shipper.ship_special(f"PermissionError executing {cmd}", metadata, emit_stdout_logs)
+        # 126 is the standard bash exit code for permission failure.
+        return 126
+    except Exception:
+        msg = f"unexpected failure executing {cmd}:\n" + traceback.format_exc()
+        shipper.ship_special(msg, metadata, emit_stdout_logs)
+        # 80 is the exit code we use to signal "ship_logs.py failed"
+        return 80
+
+    # Just for mypy.
+    assert isinstance(p.stdout, io.RawIOBase) and isinstance(p.stderr, io.RawIOBase)
+
+    try:
+        stdout = Collector(p.stdout, "stdout", emit_stdout_logs, metadata, logq, doneq)
+        stderr = Collector(p.stderr, "stderr", emit_stdout_logs, metadata, logq, doneq)
+        waiter = threading.Thread(target=lambda: doneq.put(("waiter", p.wait(), None)))
+
+        waiter.start()
+        waiter_started = True
+
+        # Set up signal forwarding.
+        def signal_passthru(signum: Any, frame: Any):
+            p.send_signal(signum)
+
+        for sig in [
+            signal.SIGINT,
+            signal.SIGTERM,
+            signal.SIGHUP,
+            signal.SIGUSR1,
+            signal.SIGUSR2,
+            signal.SIGWINCH,
+        ]:
+            signal.signal(sig, signal_passthru)
+
+        stdout.start()
+        stdout_started = True
+
+        stderr.start()
+        stderr_started = True
+
+        shipper.start()
+        shipper_started = True
+
+        # expect 4 messages on the doneq
+        first_who = None
+        first_error = None
+        exit_code = None
+        deadline = None  # type: Optional[float]
+        for _ in range(4):
+            # Wait for an event, possibly with a deadline (if the child process already exited).
+            try:
+                timeout = None if deadline is None else deadline - time.time()
+                if timeout is not None and timeout <= 0:
```

Comment:
If the deadline passed in-between calls to `doneq.get()`, since the deadline is a fixed point in time that could be passed by any given line of code.

Unfortunately, `q.get(timeout=-1)` isn't interpreted as "oh your deadline already passed" it's a ValueError.  And `q.get(timeout=0)` isn't interpreted as "oh your deadline is right now" it means "block forever".

That decision makes sense from the standard library because it's fairly common to see `timeout=-1` or `timeout=0` as ways to specify "block forever" kind of behavior, especially in C where you might use `int` or `unsigned int` for your timeout (no `None` value available).

But this decision is also annoying because this limitation comes up every time you want this "pop until deadline" kind of behavior.